### ISSUE TITLE
Listing and Downloading models file recursively

### ISSUE THREAD

**Issue by hassanzadeh** [huggingface/huggingface_hub#1969]
### Feature request

Hey Guys,
Is there an http-like api that allows to recursively download all model files? I know there is an API that you can you with python. But I need to do that in my iOS/macOS app (through swift).

Thanks

### Motivation

for android and iOS app the current API is not usable.

### Your contribution

N/A

**Comment by ArthurZucker** [comment]
cc @Wauplin might have an idea, moving this to the `huggingface_hub` library it's not a bug rather a question

**Comment by Wauplin** [comment]
Hi @hassanzadeh, in Python it's implemented in [snapshot_download.py](https://github.com/huggingface/huggingface_hub/blob/cca2882634fd2b0fa50671cac8ed793753697f44/src/huggingface_hub/_snapshot_download.py#L35) if you want to take inspiration from it. What we do is:
1. Call an endpoint that gives information about a model, including all of its files. Example: https://huggingface.co/api/models/microsoft/phi-2 (see the `"siblings"` key).
2. For each file, we download it using the `/resolve` endpoint. Example: https://huggingface.co/microsoft/phi-2/resolve/main/model-00001-of-00002.safetensors.

The logic in Python is a bit more complex since we are caching the files. Before downloading a file, we make sure it is not already in the cache. The cache system is based on the "etag" associated to each file on the Hub (etag is unique per repo, not across the Hub). If you want to learn more on how we've built the cache in the Python client, check out [this guide](https://huggingface.co/docs/huggingface_hub/guides/manage-cache). You might not need this in your case.

Hope these links will help you integrate your HF models to your app :) Please let me know if you have further questions, otherwise you can close the issue :hugs:

**Comment by hassanzadeh** [comment]
Hi @Wauplin  That was helpful, thank you :)

**Comment by julien-c** [comment]
note that @pcuenca might have implemented some hub-download code in Swift in the past (if i'm not mistaken)

**Comment by pcuenca** [comment]
Yes, I wrote a [snapshot downloader in Swift](https://github.com/huggingface/swift-transformers/blob/cb0ce3f3783dea2f3469cf3bfa6d179ea2070fc3/Sources/Hub/HubApi.swift#L174) with some of the features of the Python API (authentication, caching, allow patterns), but not all. It's a module in the [`swift-transformers` package](https://github.com/huggingface/swift-transformers). Here's a [command-line tool](https://github.com/huggingface/swift-transformers/blob/cb0ce3f3783dea2f3469cf3bfa6d179ea2070fc3/Sources/HubCLI/HubCLI.swift#L58) that shows how to use it, and [additional tests](https://github.com/huggingface/swift-transformers/blob/main/Tests/HubTests/HubApiTests.swift) that exercise the API.

@hassanzadeh Please, let me know if that helps or if you encounter any issues.

**Comment by hassanzadeh** [comment]
Perfect, thank you @pcuenca :)

**Comment by hassanzadeh** [comment]
By the way, does the download module check the SHA key of binary files just to make sure the user did download the file completely?

**Comment by Wauplin** [comment]
> By the way, does the download module check the SHA key of binary files just to make sure the user did download the file completely?

At least for the Python client, we are not doing it as it would incur additional compute. We only check the size of the downloaded file and assume that if size is consistent then "it's fine".

**Comment by pcuenca** [comment]
We don't do it in the Swift package either. We don't even verify the downloaded size yet, that could be a nice improvement to make.
